torchtitan/experiments/deepseek_v3/model.py

```diff
@@ -706,23 +730,156 @@ def moe_forward(self, x, topk_ids, topk_weight):
         # Since this is an "aritificial" index creation (final outcome being
         # `idxs`), we don't need gradients here.
+
         with torch.no_grad():
             # [seq_len, n_routed_experts]
-            cnts = topk_ids.new_zeros((topk_ids.shape[0], self.config.n_routed_experts))
+            expert_counts = topk_ids.new_zeros(
+                (topk_ids.shape[0], self.config.n_routed_experts)
+            )
             # Fill 1 to the selected experts
-            cnts.scatter_(1, topk_ids, 1)
-            tokens_per_expert = cnts.sum(dim=0)
+            expert_counts.scatter_(1, topk_ids, 1)
+            tokens_per_expert = expert_counts.sum(dim=0)
             # Token indices for each expert
-            idxs = topk_ids.view(-1).argsort()
-            sorted_tokens_shape = idxs.shape + x.shape[1:]
+            token_indices = topk_ids.view(-1).argsort()
+
+            sorted_tokens_shape = token_indices.shape + x.shape[1:]
+
+        sorted_tokens = x[token_indices // topk_ids.shape[1]]
+        # assert sorted_tokens.shape == sorted_tokens_shape
+
+        return (sorted_tokens, token_indices, tokens_per_expert)
+
+    # ------- Group GEMM implementation ------
+    def _run_group_gemm(self, contig_tokens, m_sizes, m_offsets, permuted_indices):
+        if self.group_mm in ["torch", "torchao"]:
+            # Get weights
+            w_gate = self.get_parameter("gate_proj_weight")
+            w_up = self.get_parameter("up_proj_weight")
+            w_down = self.get_parameter("down_proj_weight")
+
+            # Run first two GEMMs (gate and up projections)
+            if self.group_mm == "torchao":
+                gate_proj = grouped_gemm_forward(contig_tokens, w_gate, m_sizes)
+                up_proj = grouped_gemm_forward(contig_tokens, w_up, m_sizes)
+            else:  # "torch"
+                gate_proj = torch._grouped_mm(
+                    contig_tokens,
+                    w_gate.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+                up_proj = torch._grouped_mm(
+                    contig_tokens,
+                    w_up.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+
+            # Apply activation
+            hidden_outputs = MLP.act_fn(gate_proj) * up_proj
+
+            # Run the third GEMM (down projection)
+            if self.group_mm == "torchao":
+                hidden_outputs = grouped_gemm_forward(hidden_outputs, w_down, m_sizes)
+            else:  # "torch"
+                hidden_outputs = torch._grouped_mm(
+                    hidden_outputs,
+                    w_down.transpose(-2, -1),
+                    m_offsets,
+                    out_dtype=torch.bfloat16,
+                )
+
+            return hidden_outputs
+
+        elif self.group_mm == "ds":
```

Comment:
Maybe it is worth having a map from str to function:
```
group_mlp_impls: Dict[str, Callable] = {
    "torch": group_mlp_torch,
    "torchao": group_mlp_torchao,
    "ds": group_mlp_ds,
}
```

Reply:
agree, but let me flag this as a todo b/c I think when I move into training (next) then I will add this.  This is b/c we'll have yet one more option, which is torch_scaled_group_mm (fp8 rowwise), though it requires compile. 
Anyway, adding todo for this.